### PR TITLE
make d{motion} and dw work more consistent

### PR DESCRIPTION
**Pull request by chrisbra**
Problem:  d{motion} may work different than expected (also cw
          is special cased by POSIX)
Solution: add new cpo-z flag to make the behaviour configurable

There are two special vi compatible behaviours:

1): cw behaves differently than dw. That is, because cw is special cased
    by Vim and is effectively aliased to ce.

    POSIX behaviour:
https://pubs.opengroup.org/onlinepubs/9699919799/utilities/vi.html#tag_20_152_13_81

2): d{motion} may make the whole delete operation linewise, if the start
    and end of the motion are on different lines and there are only
    blanks before the start and after the end of the motion.
    Did not find a related POSIX link that requires this behaviour.

Both behaviours can be considered inconsistent, but we cannot easily change it, because it would be a backward incompatible change and also incompatible to how classic vi behaved.

So let's add the new cpo flag "w", which when not included fixes both behaviours and make them more consistent to what users would expect.

This has been requested several times:
https://groups.google.com/d/msg/vim_use/aaBqT6ECkA4/ALf4odKzEDgJ https://groups.google.com/d/msg/vim_dev/Dpn3xtUF16I/T6JcOPKN6usJ http://www.reddit.com/r/vim/comments/26nut8/why_does_cw_work_like_ce/ https://groups.google.com/d/msg/vim_use/vunNWLFWfQg/MmJh_ZGaAgAJ https://github.com/vim/vim/issues/4390

So in summary, if you want to have the w motion work more consistent, remove the 'z' from the cpo settings.

related: #4390

### CODE REVIEW

**Comment by zeertzjq** [comment]
> So let's add the new cpo flag "w", which when not included fixes both

This should be "z" in the commit message now.

File: src/testdir/test_cpoptions.vim [hunk 1]
```diff
@@ -912,4 +912,50 @@ func Test_cpo_dot()
   let &cpo = save_cpo
 endfunc
 
+" Test for the 'z' flag in 'cpo' (make cw and dw work similar and avoid
+" inconsistencies, see :h cpo-z)
+" modified)
```
Comment (zeertzjq) [inline src/testdir/test_cpoptions.vim]: This line looks strange

File: runtime/doc/change.txt [hunk 2]
```diff
@@ -95,13 +95,13 @@ For inserting text see |insert.txt|.
...
 |registers| for an explanation of registers.
-
+							*d-special*
 An exception for the d{motion} command: If the motion is not linewise, the
 start and end of the motion are not in the same line, and there are only
 blanks before the start and there are no non-blanks after the end of the
 motion, the delete becomes linewise.  This means that the delete also removes
 the line of blanks that you might expect to remain. Use the |o_v| operator to
-force the motion to be characterwise.
+force the motion to be characterwise or set |cpo-z|.
```
Comment (zeertzjq) [inline runtime/doc/change.txt]: `remove |cpo-z|`?

**Review by yegappan — COMMENTED** [review]
Can you mention this change in the version9.txt file under the version-9.2 section?

**Comment by Shane-XB-Qian** [comment]
BTW: welcome back~ @yegappan :smile:

--
shane.xb.qian

**Comment by yegappan** [comment]
> BTW: welcome back~ @yegappan 😄
> […](#)
> -- shane.xb.qian

Thanks.  I was out on vacation.

**Comment by chrisbra** [comment]
> Can you mention this change in the version9.txt file under the version-9.2 section?

Yes planned for when I merge it.
